core/editable.js

```diff
@@ -1717,14 +1718,24 @@
 			// Split inline elements so HTML will be inserted with its own styles.
 			path = range.startPath();
 			if ( ( node = path.contains( isInline, false, 1 ) ) ) {
-				range.splitElement( node );
+				clone = range.splitElement( node );
 				that.inlineStylesRoot = node;
 				that.inlineStylesPeak = path.lastElement;
 			}
 
 			// Record inline merging candidates for later cleanup in place.
 			bm = range.createBookmark();
 
+			// When called by insertHtml remove empty element created after splitting (#2813).
+			if ( isHtml && clone && clone.isEmptyInlineRemoveable() ) {
```

Comment:
Would be nice to extract both `span` node removals into a separate function to avoid duplication.